File: iEEGsc.m
classdef iEEGsc

    properties
        whole_brain_trk
        whole_brain_trksubVox
        trk_to_t1surfRAS
        freeSurferDir
        output
        IEEGdata
    end

    methods

        function electrodes = iEEGgrey2white(obj)

            outdir = fullfile(obj.output, 'connectivityIEEG');
            mkdir(outdir)

            try
                mustBeFile([outdir '/electrodes_surf_proj.mat'])
                load([outdir '/electrodes_surf_proj.mat'],'electrodes');

            catch


                electrodes = readtable(obj.IEEGdata);

                % remove the electodes outside the brain
                outside = ismember(electrodes.roi,'outside-brain');
                electrodes(outside,:) = [];

                % Find contacts in the grey matter contex only
                in_cortex =  electrodes.roiNum>999;
                in_cortexID = find(in_cortex);

                electrodes_surf = [electrodes.surfmm_x(in_cortex), electrodes.surfmm_y(in_cortex), electrodes.surfmm_z(in_cortex)];

                [lpv, lpf] = read_surf([obj.freeSurferDir '/surf/lh.pial']);
                [lwv, lwf] = read_surf([obj.freeSurferDir '/surf/lh.white']);

                [rpv, rpf] = read_surf([obj.freeSurferDir '/surf/rh.pial']);
                [rwv, rwf] = read_surf([obj.freeSurferDir '/surf/rh.white']);

                % find if elecs are outside of WM surface
                lWM.faces=lwf+1;
                lWM.vertices=lwv;
                in_left = inpolyhedron(lWM,electrodes_surf);

                rWM.faces=rwf+1;
                rWM.vertices=rwv;
                in_right = inpolyhedron(rWM,electrodes_surf);

                out=~in_left & ~in_right;
                out_id=find(out);

                %for the electrodes outside WM, find nearest GM surface point
                vids=zeros(length(out_id),1);
                sides=cell(length(out_id),1);
                for k=1:length(out_id)
                    [mindl_id,mindl] = knnsearch(lpv,electrodes_surf(out_id(k),:));
                    [mindr_id,mindr] = knnsearch(rpv,electrodes_surf(out_id(k),:));

                    if mindl<mindr
                        vids(k)=mindl_id;
                        sides{k}='l';
                    else
                        vids(k)=mindr_id;
                        sides{k}='r';
                    end
                end

                % build up all the output variables:
                l=size(electrodes_surf,1);
                elecs_type=cell(l,1);
                elecs_side=cell(l,1);
                elecs_GMSurfxyz=zeros(l,3);
                elecs_WMSurfxyz=zeros(l,3);
                for k=1:l
                    kid=find(out_id==k);
                    if length(kid)==1%if is outside of WM
                        elecs_type{k}='OutsideWM';
                        elecs_side{k}=sides{kid};
                        if sides{kid}=='l'
                            elecs_GMSurfxyz(k,:)=lpv(vids(kid),:);
                            elecs_WMSurfxyz(k,:)=lwv(vids(kid),:);
                        else
                            elecs_GMSurfxyz(k,:)=rpv(vids(kid),:);
                            elecs_WMSurfxyz(k,:)=rwv(vids(kid),:);
                        end
                    else % if inside of WM
                        elecs_type{k}='InsideWM';
                        if in_left(k)
                            elecs_side{k}='l';

                        else
                            elecs_side{k}='r';
                        end
                        elecs_GMSurfxyz(k,:)=[NaN NaN NaN];
                        elecs_WMSurfxyz(k,:)=[NaN NaN NaN];
                    end
                end

                in_cortexIDGM = in_cortexID(out_id);

                electrodes.surfmm_x(in_cortexIDGM) = elecs_WMSurfxyz(out_id,1);
                electrodes.surfmm_y(in_cortexIDGM) = elecs_WMSurfxyz(out_id,2);
                electrodes.surfmm_z(in_cortexIDGM) = elecs_WMSurfxyz(out_id,3);

                save([outdir '/electrodes_surf_proj.mat'],"electrodes")
            end

        end


        function edgeList = makeEdgeList(obj,electrodes,sphereDia,varargin)

            outdir = fullfile(obj.output, 'connectivityIEEG');
            mkdir(outdir)

            try

                mustBeFile([outdir '/edgeList_' num2str(sphereDia) 'mmSph.mat'])
                load([outdir '/edgeList_' num2str(sphereDia) 'mmSph.mat'],"edgeList");

            catch

                iEEGprojected = [electrodes.surfmm_x, electrodes.surfmm_y, electrodes.surfmm_z];
                edgeList = [];

                load(obj.whole_brain_trk,"trk");
                fieldsToKeep = {'length','cord','startEnd'};
                trk = rmfield(trk,setdiff(fieldnames(trk), fieldsToKeep));
                xform.trk_to_t1surfRAS = readmatrix(obj.trk_to_t1surfRAS);
                trk.cord = xform.trk_to_t1surfRAS * trk.cord; % Transform tracts to surface space

                load(obj.whole_brain_trksubVox,"trksubVox");
                trk.subVoxQA = trksubVox.qa;
                trk.subVoxFA = trksubVox.fa;
                trk.subVoxMD = trksubVox.md;
                trk.subVoxAD = trksubVox.ad;
                trk.subVoxRD = trksubVox.rd;
                clear trksubVox

                % this is only to speed up by diving the data in small bins
                nBins = transpose(histcounts(1:numel(trk.length),1000));
                tBinEnd= cumsum(nBins);
                tBinStart=[1; 1+tBinEnd];
                tBinStart = tBinStart(1:end-1);
                trkPosition = [0;tBinEnd(1:end-1)];
                for tBin = 1:numel(tBinStart)
                    trkBin.length = trk.length(tBinStart(tBin):tBinEnd(tBin));

                    trkBinEnds=(cumsum(double(trkBin.length)));
                    trkBinStarts=[1; 1+trkBinEnds];
                    trkBin.startEnd = [trkBinStarts(1:end-1),trkBinEnds];

                    trkBin.cord = trk.cord(:,trk.startEnd(tBinStart(tBin),1):trk.startEnd(tBinEnd(tBin),2));
                    trkBin.subVoxQA = trk.subVoxQA(:,trk.startEnd(tBinStart(tBin),1):trk.startEnd(tBinEnd(tBin),2));
                    trkBin.subVoxFA = trk.subVoxFA(:,trk.startEnd(tBinStart(tBin),1):trk.startEnd(tBinEnd(tBin),2));
                    trkBin.subVoxMD = trk.subVoxMD(:,trk.startEnd(tBinStart(tBin),1):trk.startEnd(tBinEnd(tBin),2));
                    trkBin.subVoxAD = trk.subVoxAD(:,trk.startEnd(tBinStart(tBin),1):trk.startEnd(tBinEnd(tBin),2));
                    trkBin.subVoxRD = trk.subVoxRD(:,trk.startEnd(tBinStart(tBin),1):trk.startEnd(tBinEnd(tBin),2));

                    trkBin.tPosition = trkPosition(tBin);

                    % do par for here
                    parfor t=1:numel(trkBin.length)
                        [~,d] = knnsearch(trkBin.cord(1:3,trkBin.startEnd(t,1):trkBin.startEnd(t,2))',iEEGprojected);

                        if sum(d<sphereDia) > 1

                            connectedElecs = find(d <= sphereDia);
                            trkProp = iEEGsc.getTrkMetricsPass(trkBin,t,iEEGprojected,connectedElecs,sphereDia);
                            edgeList = [edgeList;trkProp];

                        end
                    end
                    clear trkBin
                end

                edgeList = array2table(edgeList,'VariableNames',{'roi1','roi2','length',...
                    'qa','fa','md','ad','rd','trkindx'});

                save([outdir '/edgeList_' num2str(sphereDia) 'mmSph.mat'],"edgeList");
            end

        end

        function connectivity = makeConnectivityMatrix(obj,edgeList,sphereDia)

            outdir = fullfile(obj.output, 'connectivityIEEG');
            mkdir(outdir)

            try

                mustBeFile([outdir '/connectivity_' num2str(sphereDia) 'mmSph.mat']);
                load([outdir '/connectivity_' num2str(sphereDia) 'mmSph.mat'],"connectivity")
            
            catch

                electrodes = readtable(obj.IEEGdata);
                % remove the electodes outside the brain
                outside = ismember(electrodes.roi,'outside-brain');
                electrodes(outside,:) = [];

                connectivity.count=zeros(size(electrodes,1));
                connectivity.len=zeros(size(electrodes,1));
                connectivity.qa=zeros(size(electrodes,1));
                connectivity.fa=zeros(size(electrodes,1));
                connectivity.md=zeros(size(electrodes,1));
                connectivity.ad=zeros(size(electrodes,1));
                connectivity.rd=zeros(size(electrodes,1));

                for i=1:size(edgeList,1)

                    roi1 = edgeList.roi1(i);
                    roi2 = edgeList.roi2(i);

                    connectivity.count(roi1,roi2) = connectivity.count(roi1,roi2)+1;
                    connectivity.len(roi1,roi2)= connectivity.len(roi1,roi2)+ edgeList.length(i);
                    connectivity.qa(roi1,roi2)=connectivity.qa(roi1,roi2) + edgeList.qa(i);
                    connectivity.fa(roi1,roi2)=connectivity.fa(roi1,roi2) + edgeList.fa(i);
                    connectivity.md(roi1,roi2)=connectivity.md(roi1,roi2) + edgeList.md(i);
                    connectivity.ad(roi1,roi2)=connectivity.ad(roi1,roi2) + edgeList.ad(i);
                    connectivity.rd(roi1,roi2)=connectivity.rd(roi1,roi2) + edgeList.rd(i);

                end

                %% Make adjacency matrix symmetric
                connectivity.count = connectivity.count'+connectivity.count;
                connectivity.len = connectivity.len' + connectivity.len;
                connectivity.qa = connectivity.qa' + connectivity.qa;
                connectivity.fa = connectivity.fa' + connectivity.fa;
                connectivity.md = connectivity.md' + connectivity.md;
                connectivity.ad = connectivity.ad' + connectivity.ad;
                connectivity.rd = connectivity.rd' + connectivity.rd;

                %% Take the mean of length, qa, gfa, md, fa
                connectivity.len = connectivity.len./connectivity.count;
                connectivity.qa = connectivity.qa./connectivity.count;
                connectivity.fa = connectivity.fa./connectivity.count;
                connectivity.md = connectivity.md./connectivity.count;
                connectivity.ad = connectivity.ad./connectivity.count;
                connectivity.rd = connectivity.rd./connectivity.count;

                %% Half the diagonal of count
                connectivity.count(1:size(connectivity.count,1)+1:end)=connectivity.count(1:size(connectivity.count,1)+1:end)/2;

                connectivity.len(isnan(connectivity.len)) = 0;
                connectivity.qa(isnan(connectivity.qa)) = 0;
                connectivity.fa(isnan(connectivity.fa)) = 0;
                connectivity.md(isnan(connectivity.md)) = 0;
                connectivity.ad(isnan(connectivity.ad)) = 0;
                connectivity.rd(isnan(connectivity.rd)) = 0;

                save([outdir '/connectivity_' num2str(sphereDia) 'mmSph.mat'],"connectivity","electrodes");

            end
        end

    end



    methods (Static)  % Notice the Static attribute here

        function [trkProp,cordTrk] = getTrkMetricsPass(trk,t,roiCord,connectedElecs,sphereDist)

            trkIntrst = trk.cord(1:3,trk.startEnd(t,1):trk.startEnd(t,2))';
            elecCord = roiCord(connectedElecs,:);

            % For each electrode find the index of the nearest point on the tract
            for elec = 1:numel(connectedElecs)
                elecOnTrk.electrode(elec,:) = connectedElecs(elec);
                [~,dtrk] = knnsearch(elecCord(elec,:),trkIntrst);
                elecOnTrk.atTractPos(elec,:) = find(dtrk<=sphereDist,1);
            end

            elecOnTrk = struct2table(elecOnTrk);
            elecOnTrk = sortrows(elecOnTrk,'atTractPos','ascend');
            roiPairs = combnk(elecOnTrk.electrode,2);
            cordTrk = [];
            for roiList = 1:size(roiPairs,1)

                posStart = elecOnTrk.atTractPos(elecOnTrk.electrode == roiPairs(roiList,1));
                posEnd = elecOnTrk.atTractPos(elecOnTrk.electrode == roiPairs(roiList,2));
                startEnd = trk.startEnd(t,1):trk.startEnd(t,2);

                roi1(roiList,:) = roiPairs(roiList,1);
                roi2(roiList,:) = roiPairs(roiList,2);
                lengthTrk(roiList,:) = numel(startEnd(posStart):startEnd(posEnd));
                qaTrk(roiList,:) = mean(trk.subVoxQA(startEnd(posStart):startEnd(posEnd)));
                faTrk(roiList,:) = mean(trk.subVoxFA(startEnd(posStart):startEnd(posEnd)));
                mdTrk(roiList,:) = mean(trk.subVoxMD(startEnd(posStart):startEnd(posEnd)));
                adTrk(roiList,:) = mean(trk.subVoxAD(startEnd(posStart):startEnd(posEnd)));
                rdTrk(roiList,:) = mean(trk.subVoxRD(startEnd(posStart):startEnd(posEnd)));


                cordTrk = [cordTrk; trkIntrst(posStart:posEnd,:)];
            end

            trkProp = [roi1,roi2,lengthTrk,qaTrk,faTrk,mdTrk,adTrk,rdTrk,repmat(t+trk.tPosition,size(roiPairs,1),1)];

        end

    end
end

File: dwi_preprocessing/ieeg_sc.py
"""IEEGsc — Python port of iEEGsc.m

Computes structural connectivity between iEEG electrode contacts using
whole-brain fiber tracts and per-point diffusion metrics (QA, FA, MD, AD, RD).
"""

import os
from itertools import combinations
from pathlib import Path

import numpy as np
import pandas as pd
from scipy.spatial import KDTree
from joblib import Parallel, delayed

from dwi_preprocessing.utils.surfaces import read_surf
from dwi_preprocessing.utils.geometry import inpolyhedron
from dwi_preprocessing.utils.io import load_mat, save_mat_v73


class IEEGsc:
    """iEEG structural connectivity from DWI tractography.

    Expects the standard derivatives directory layout:
        <output>/preprocessDWI/dsiStudio/whole_brain_trk.mat
        <output>/preprocessDWI/dsiStudio/whole_brain_trksubVox.mat
        <output>/connectivityDWI/tracts_to_T1/trk_to_t1surfRAS.txt
        <output>/freesurfer/surf/{lh,rh}.{pial,white}
        <output>/ieeg_recon/module3/electrodes2ROI.csv
    """

    # Relative paths within the derivatives directory
    _PATHS = {
        "whole_brain_trk": ("preprocessDWI", "dsiStudio", "whole_brain_trk.mat"),
        "whole_brain_trksubVox": ("preprocessDWI", "dsiStudio", "whole_brain_trksubVox.mat"),
        "trk_to_t1surfRAS": ("connectivityDWI", "tracts_to_T1", "trk_to_t1surfRAS.txt"),
        "freeSurferDir": ("freesurfer",),
        "IEEGdata": ("ieeg_recon", "module3", "electrodes2ROI.csv"),
    }

    def __init__(self, output: str | Path):
        self.output = Path(output)

    def _path(self, key: str) -> Path:
        return self.output / os.path.join(*self._PATHS[key])

    # ------------------------------------------------------------------
    # Step 1: Project grey-matter electrodes to white-matter surface
    # ------------------------------------------------------------------

    def ieeg_grey2white(self) -> pd.DataFrame:
        """Project cortical (grey-matter) electrodes onto the WM surface.

        Returns a DataFrame with electrode positions updated so that
        contacts originally in grey matter are projected to the nearest
        white-matter surface vertex.
        """
        outdir = self.output / "connectivityIEEG"
        outdir.mkdir(exist_ok=True)
        cache = outdir / "electrodes_surf_proj.mat"

        if cache.is_file():
            print("  Electrodes already projected — loading cache")
            # Load from mat and reconstruct DataFrame
            d = load_mat(str(cache))
            return pd.DataFrame(d["electrodes"]) if "electrodes" in d else pd.read_csv(str(self._path("IEEGdata")))

        electrodes = pd.read_csv(str(self._path("IEEGdata")))

        # Remove outside-brain contacts
        electrodes = electrodes[electrodes["roi"] != "outside-brain"].reset_index(drop=True)

        # Cortical contacts only (roiNum > 999)
        in_cortex = electrodes["roiNum"] > 999
        in_cortex_idx = np.where(in_cortex)[0]

        elec_surf = electrodes.loc[in_cortex, ["surfmm_x", "surfmm_y", "surfmm_z"]].values

        # Load surfaces
        fs_dir = str(self._path("freeSurferDir"))
        lpv, lpf = read_surf(os.path.join(fs_dir, "surf", "lh.pial"))
        lwv, lwf = read_surf(os.path.join(fs_dir, "surf", "lh.white"))
        rpv, rpf = read_surf(os.path.join(fs_dir, "surf", "rh.pial"))
        rwv, rwf = read_surf(os.path.join(fs_dir, "surf", "rh.white"))

        # Test which electrodes are inside the WM surfaces
        in_left = inpolyhedron(lwv, lwf, elec_surf)
        in_right = inpolyhedron(rwv, rwf, elec_surf)
        outside_wm = ~in_left & ~in_right
        out_ids = np.where(outside_wm)[0]

        # For electrodes outside WM, find nearest pial vertex
        lp_tree = KDTree(lpv)
        rp_tree = KDTree(rpv)

        vids = np.zeros(len(out_ids), dtype=int)
        sides = [""] * len(out_ids)

        for k, oid in enumerate(out_ids):
            dl, il = lp_tree.query(elec_surf[oid])
            dr, ir = rp_tree.query(elec_surf[oid])
            if dl < dr:
                vids[k] = il
                sides[k] = "l"
            else:
                vids[k] = ir
                sides[k] = "r"

        # Build WM projection coordinates
        wm_xyz = np.full((len(out_ids), 3), np.nan)
        for k in range(len(out_ids)):
            if sides[k] == "l":
                wm_xyz[k] = lwv[vids[k]]
            else:
                wm_xyz[k] = rwv[vids[k]]

        # Update electrode coordinates: project GM contacts to WM surface
        gm_global_idx = in_cortex_idx[out_ids]
        electrodes.loc[gm_global_idx, "surfmm_x"] = wm_xyz[:, 0]
        electrodes.loc[gm_global_idx, "surfmm_y"] = wm_xyz[:, 1]
        electrodes.loc[gm_global_idx, "surfmm_z"] = wm_xyz[:, 2]

        # Save cache as CSV (more portable than .mat for DataFrames)
        electrodes.to_csv(str(outdir / "electrodes_surf_proj.csv"), index=False)
        # Also save .mat for MATLAB compatibility
        save_mat_v73(str(cache), {"electrodes_csv_path": str(outdir / "electrodes_surf_proj.csv")})

        return electrodes

    # ------------------------------------------------------------------
    # Step 2: Build edge list (tract-electrode connections)
    # ------------------------------------------------------------------

    def make_edge_list(self, electrodes: pd.DataFrame,
                       sphere_dia: float) -> pd.DataFrame:
        """Find tracts passing within sphere_dia of electrode pairs.

        Returns a DataFrame with columns:
            roi1, roi2, length, qa, fa, md, ad, rd, trkindx
        """
        outdir = self.output / "connectivityIEEG"
        outdir.mkdir(exist_ok=True)
        cache = outdir / f"edgeList_{sphere_dia}mmSph.csv"

        if cache.is_file():
            print(f"  Edge list ({sphere_dia}mm) already exists — loading")
            return pd.read_csv(str(cache))

        ieeg_projected = electrodes[["surfmm_x", "surfmm_y", "surfmm_z"]].values

        # Load tracts
        print(f"  Loading whole_brain_trk.mat...")
        trk_raw = load_mat(str(self._path("whole_brain_trk")))
        # Handle nested h5py struct
        if "trk" in trk_raw and isinstance(trk_raw["trk"], dict):
            trk = trk_raw["trk"]
        else:
            trk = trk_raw

        cord = trk["cord"]  # (4, N_points)
        length = trk["length"].ravel()
        start_end = trk["startEnd"]  # (N_tracts, 2) — 1-indexed from MATLAB

        # Load transform and apply
        xform = np.loadtxt(str(self._path("trk_to_t1surfRAS")))
        cord = xform @ cord  # Transform to surface RAS

        # Load subvoxel metrics
        print(f"  Loading whole_brain_trksubVox.mat...")
        subvox_raw = load_mat(str(self._path("whole_brain_trksubVox")))
        if "trksubVox" in subvox_raw and isinstance(subvox_raw["trksubVox"], dict):
            sv = subvox_raw["trksubVox"]
        else:
            sv = subvox_raw

        subvox_qa = sv["qa"].ravel()
        subvox_fa = sv["fa"].ravel()
        subvox_md = sv["md"].ravel()
        subvox_ad = sv["ad"].ravel()
        subvox_rd = sv["rd"].ravel()

        # Convert start_end to 0-indexed Python
        se = start_end.astype(np.int64)
        se[:, 0] -= 1  # 1-indexed -> 0-indexed start
        se[:, 1] -= 1  # 1-indexed -> 0-indexed end (inclusive)

        # Process in bins for memory efficiency (same as MATLAB)
        n_tracts = len(length)
        n_bins = 1000
        bin_edges = np.linspace(0, n_tracts, n_bins + 1, dtype=int)

        cord_xyz = cord[:3, :]  # (3, N_points)

        print(f"  Building edge list ({sphere_dia}mm) over {n_tracts} tracts...")
        all_edges = []

        for b in range(n_bins):
            b_start = bin_edges[b]
            b_end = bin_edges[b + 1]
            if b_start >= b_end:
                continue

            bin_results = Parallel(n_jobs=-1, prefer="threads")(
                delayed(self._process_tract)(
                    t, cord_xyz, se, subvox_qa, subvox_fa, subvox_md,
                    subvox_ad, subvox_rd, ieeg_projected, sphere_dia, b_start
                )
                for t in range(b_start, b_end)
            )

            for res in bin_results:
                if res is not None:
                    all_edges.append(res)

        if all_edges:
            edge_arr = np.vstack(all_edges)
            edge_list = pd.DataFrame(edge_arr, columns=[
                "roi1", "roi2", "length", "qa", "fa", "md", "ad", "rd", "trkindx"
            ])
        else:
            edge_list = pd.DataFrame(columns=[
                "roi1", "roi2", "length", "qa", "fa", "md", "ad", "rd", "trkindx"
            ])

        edge_list.to_csv(str(cache), index=False)
        # Also save .mat for MATLAB compatibility
        save_mat_v73(
            str(outdir / f"edgeList_{sphere_dia}mmSph.mat"),
            {"edgeList": edge_arr if all_edges else np.array([])}
        )
        return edge_list

    @staticmethod
    def _process_tract(t, cord_xyz, se, sq, sf, sm, sa, sr,
                       ieeg_projected, sphere_dia, offset):
        """Process a single tract: find electrodes within sphere_dia."""
        s, e = int(se[t, 0]), int(se[t, 1]) + 1  # inclusive -> exclusive
        tract_pts = cord_xyz[:, s:e].T  # (n_pts, 3)

        tree = KDTree(tract_pts)
        dists, _ = tree.query(ieeg_projected)

        connected = np.where(dists <= sphere_dia)[0]
        if len(connected) < 2:
            return None

        # For each connected electrode, find its position along the tract
        elec_on_trk = []
        for elec_idx in connected:
            d_to_tract = np.linalg.norm(tract_pts - ieeg_projected[elec_idx], axis=1)
            pos = np.argmin(d_to_tract)
            if d_to_tract[pos] <= sphere_dia:
                elec_on_trk.append((elec_idx, pos))

        if len(elec_on_trk) < 2:
            return None

        # Sort by position along tract
        elec_on_trk.sort(key=lambda x: x[1])
        elec_indices = [x[0] for x in elec_on_trk]
        elec_positions = [x[1] for x in elec_on_trk]

        # All pairs
        results = []
        for (i, ei), (j, ej) in combinations(range(len(elec_indices)), 2):
            roi1 = elec_indices[i] + 1  # 1-indexed for MATLAB compat
            roi2 = elec_indices[j] + 1
            pos_start = elec_positions[i]
            pos_end = elec_positions[j]

            abs_start = s + pos_start
            abs_end = s + pos_end + 1  # exclusive

            seg_len = pos_end - pos_start + 1
            seg_qa = np.mean(sq[abs_start:abs_end])
            seg_fa = np.mean(sf[abs_start:abs_end])
            seg_md = np.mean(sm[abs_start:abs_end])
            seg_ad = np.mean(sa[abs_start:abs_end])
            seg_rd = np.mean(sr[abs_start:abs_end])

            results.append([roi1, roi2, seg_len, seg_qa, seg_fa,
                            seg_md, seg_ad, seg_rd, t + 1])  # 1-indexed tract

        return np.array(results)

    # ------------------------------------------------------------------
    # Step 3: Build connectivity matrices
    # ------------------------------------------------------------------

    def make_connectivity_matrix(self, edge_list: pd.DataFrame,
                                 sphere_dia: float) -> dict:
        """Aggregate edge list into symmetric connectivity matrices."""
        outdir = self.output / "connectivityIEEG"
        outdir.mkdir(exist_ok=True)
        cache = outdir / f"connectivity_{sphere_dia}mmSph.mat"

        if cache.is_file():
            print(f"  Connectivity matrix ({sphere_dia}mm) already exists — loading")
            return load_mat(str(cache))

        # Load electrode count
        electrodes = pd.read_csv(str(self._path("IEEGdata")))
        electrodes = electrodes[electrodes["roi"] != "outside-brain"].reset_index(drop=True)
        n = len(electrodes)

        metrics = ["count", "len", "qa", "fa", "md", "ad", "rd"]
        conn = {m: np.zeros((n, n)) for m in metrics}

        for _, row in edge_list.iterrows():
            r1 = int(row["roi1"]) - 1  # Back to 0-indexed
            r2 = int(row["roi2"]) - 1
            if r1 >= n or r2 >= n:
                continue
            conn["count"][r1, r2] += 1
            conn["len"][r1, r2] += row["length"]
            conn["qa"][r1, r2] += row["qa"]
            conn["fa"][r1, r2] += row["fa"]
            conn["md"][r1, r2] += row["md"]
            conn["ad"][r1, r2] += row["ad"]
            conn["rd"][r1, r2] += row["rd"]

        # Symmetrise
        for m in metrics:
            conn[m] = conn[m] + conn[m].T

        # Mean (divide by count)
        for m in ["len", "qa", "fa", "md", "ad", "rd"]:
            with np.errstate(divide="ignore", invalid="ignore"):
                conn[m] = np.where(conn["count"] > 0,
                                   conn[m] / conn["count"], 0.0)

        # Half the diagonal of count
        np.fill_diagonal(conn["count"], np.diag(conn["count"]) / 2)

        # Replace NaN with 0
        for m in metrics:
            conn[m] = np.nan_to_num(conn[m], nan=0.0)

        save_mat_v73(str(cache), {"connectivity": conn})
        # Also save electrode info
        electrodes.to_csv(str(outdir / f"electrodes_{sphere_dia}mmSph.csv"), index=False)

        print(f"  Connectivity matrix ({sphere_dia}mm) saved: {n}x{n}")
        return conn
